path_rotate_root_test.go

```diff
@@ -0,0 +1,196 @@
+package azureauth
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/vault/sdk/logical"
+)
+
+func TestRotateRootSuccess(t *testing.T) {
+	b, s := getTestBackend(t)
+
+	subscriptionID, tenantID, clientID, clientSecret := getAzureEnvironmentSettings()
+	if subscriptionID == "" || tenantID == "" ||
+		clientID == "" || clientSecret == "" {
+		t.Skip("environment variables not set, skipping test in CI")
+	}
+
+	configData := map[string]interface{}{
+		"subscription_id": subscriptionID,
+		"tenant_id":       tenantID,
+		"resource":        "https://management.azure.com/",
+		"client_id":       clientID,
+		"client_secret":   clientSecret,
+	}
+	if err := testConfigCreate(t, b, s, configData); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	resp, err := b.HandleRequest(context.Background(), &logical.Request{
+		Operation: logical.UpdateOperation,
+		Path:      "rotate-root",
+		Data:      map[string]interface{}{},
+		Storage:   s,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if resp != nil && resp.IsError() {
+		t.Fatal(resp.Error())
+	}
+
+	config, err := b.config(context.Background(), s)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if config.ClientSecret == "" {
+		t.Fatal(fmt.Errorf("root password was empty after rotate root, it shouldn't be"))
+	}
+
+	if config.NewClientSecret == config.ClientSecret {
+		t.Fatal("old and new password equal after rotate-root, it shouldn't be")
+	}
+
+	if config.NewClientSecret == "" {
+		t.Fatal("new password is empty, it shouldn't be")
+	}
+
+	if config.NewClientSecretKeyID == "" {
+		t.Fatal("new password key id is empty, it shouldn't be")
+	}
+
+	if !b.updatePassword {
+		t.Fatal("update password is false, it shouldn't be")
+	}
+
+	config.NewClientSecretCreated = config.NewClientSecretCreated.Add(-(time.Minute * 1))
+	err = b.saveConfig(context.Background(), config, s)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = b.periodicFunc(context.Background(), &logical.Request{
+		Storage: s,
+	})
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	newConfig, err := b.config(context.Background(), s)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if newConfig.ClientSecret != config.NewClientSecret {
+		t.Fatal(fmt.Errorf("old and new password aren't equal after periodic function, they should be"))
+	}
+}
+
+func TestRotateRootPeriodicFunctionBeforeMinute(t *testing.T) {
```

Comment:
Not sure of what the expected behavior of this test should be. The test passes and a new client credential is created for my azure application, however the periodic func doesn't actually remove the old application password. Would appreciate any feedback on what this test should actually do 🙏🏼

Reply:
This test seems to be verifying that we don't do a rotation until `NewClientSecretCreated` is greater than 1 minute. See https://github.com/hashicorp/vault-plugin-auth-azure/pull/88/files#diff-f7febf2126e4c13f0730b2d06b8a6d3bdcb748f6379770c478888c2e4ea1949eR96

So I think it is ok, assuming that is the behavior that we want. Maybe we could add a comment above this test describing the purpose?

Reply:
You can check it by adding this log line [before that return nil](https://github.com/hashicorp/vault-plugin-auth-azure/pull/88/files#diff-f7febf2126e4c13f0730b2d06b8a6d3bdcb748f6379770c478888c2e4ea1949eR97)

```
b.Logger().Debug("periodic func", "rotate-root", "Password should be at least a minute old before we process it")
```
and then add another test case here
```
{
	Name:    "70 seconds test:",
	Created: time.Second * 70,
},

```

Reply:
Ah nice! Thanks for the clarification; I'll add a brief description to the test, because this confused me a bit 😅

Reply:
Added a docstring to the test to better describe its working. Thanks! 🙏🏼